### PR TITLE
storage/ingest: Support for consume-from-position-at-startup=timestamp

### PR DESCRIPTION
**Pull request by narqo**
#### What this PR does

This one adds a new `ingest-storage.kafka.consume-from-position-at-startup=timestamp` flag. The flag makes the ingest store's partition reader to start consuming from a specific milliseconds timestamp (in oppose to consuming from earliest or latest offsets).

The rationale for the flag is the following:

In case of an incident where the consumer group is lost, at the next ingester startup it will start consuming the partition from start or from end. Both options aren't particularly useful to handle such scenario. Consuming from start could mean replaying a bunch of data already ingested; consuming from end may cause data loss (because the data in between hasn't been consumed).

The new option is a middle ground between two. It allows starting the ingester with a rough point in time, skipping the backlog at the head of the topic.

I've tested it manually against a local Kafka running in docker, and it works as I expected.

#### Checklist

- [x] Tests updated.
- [ ] Documentation added.
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`.
- [ ] [`about-versioning.md`](https://github.com/grafana/mimir/blob/main/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

### CODE REVIEW

**Comment by narqo** [comment]
Note, [failing tests][1] is something _unexpectedly_ "expected". It seems that kgo's fake cluster behaves differently from the real Kafka. I'll be delving into that.

[1]: https://github.com/grafana/mimir/actions/runs/8658667282/job/23743001092

File: pkg/storage/ingest/config.go [hunk 1]
```diff
@@ -98,6 +101,16 @@ func (cfg *KafkaConfig) Validate() error {
 	if !slices.Contains(consumeFromPositionOptions, cfg.ConsumeFromPositionAtStartup) {
 		return ErrInvalidConsumePosition
 	}
+	if cfg.ConsumeFromPositionAtStartup == consumeFromTimestamp {
+		// We only do a simple soundness check for the value be a millisecond precision timestamp.
+		if cfg.ConsumeFromTimestampAtStartup < 1e12 {
```
Comment (pracucci) [inline pkg/storage/ingest/config.go:106]: Nice idea!

File: pkg/storage/ingest/reader.go [hunk 2]
```diff
@@ -437,22 +408,73 @@ func (r *PartitionReader) newKafkaReader(at kgo.Offset) (*kgo.Client, error) {
...
+	// The reason for this is that changing the offset of an existing client requires to have used this client for fetching at least once.
+	// We don't want to do noop fetches just to warm up the client, so we create a new client instead.
+	cl, err := kgo.NewClient(commonKafkaClientOptions(r.kafkaCfg, r.metrics.kprom, r.logger)...)
+	if err != nil {
+		return 0, -1, fmt.Errorf("unable to create bootstrap client: %w", err)
+	}
+	defer cl.Close()
 
+	// lastConsumedOffset is unknown until (and only if) we successfully fetch its value below.
+	lastConsumedOffset = -1
```
Comment (pracucci) [inline pkg/storage/ingest/reader.go:434]: [non blocking] From a design perspective, to me it looks a bit inconsistent that `fetchOffset()` updates `lastConsumedOffset` (whose scope is outside the function) but returns the start offset. I think `fetchOffset()` could return `lastConsumedOffset` too.

File: pkg/storage/ingest/reader.go [hunk 3]
```diff
@@ -437,22 +408,73 @@ func (r *PartitionReader) newKafkaReader(at kgo.Offset) (*kgo.Client, error) {
...
+
+	fetchOffset := func(ctx context.Context) (offset int64, err error) {
+		if r.kafkaCfg.ConsumeFromPositionAtStartup == consumeFromTimestamp {
+			ts := time.UnixMilli(r.kafkaCfg.ConsumeFromTimestampAtStartup)
+			offset, exists, err := r.fetchFirstOffsetAfterTime(ctx, cl, ts)
+			if err != nil {
+				return 0, err
+			}
+			if exists {
+				level.Info(r.logger).Log("msg", "starting consumption from timestamp", "timestamp", ts.UnixMilli(), "start_offset", offset, "consumer_group", r.consumerGroup)
```
Comment (pracucci) [inline pkg/storage/ingest/reader.go:444]: I have a doubt about an edge case here. The `lastConsumedOffset` is used `processNextFetchesUntilMaxLagHonored()` to check whether the last produced offset has been consumed, and so it reader catched up. If the configured timestamp position is the end of the partition, and none is writing to the partition anymore, `processNextFetchesUntilMaxLagHonored()` may never return because `lastConsumedOffset` has not been initialised (we return -1 here) and, at the same time, it will be never set because no new records are produced. I'm wondering if, instead, we should simply initialise `lastConsumedOffset` to `offset - 1` (that is the offset before the one we'll start the consumption from).

For the other cases (consume from start/end/last-offset) this was an handled case (e.g. see the "should not wait indefinitely if there are no records to consume from Kafka but partition start offset is > 0 (e.g. all previous records have been deleted by Kafka retention)" test case, which is an edge case we've hit in dev when restarting an ingester consuming from a scaled-down partition).

**Review by pracucci — APPROVED** [review]
Excellent job! I left a couple of comments, but no blocker. I'm also happy to merge it as is, and eventually address comments in a second PR (if we'll address them).

File: pkg/storage/ingest/reader.go (hunk 3 above)
Comment (narqo) [reply pkg/storage/ingest/reader.go]: Good point, I've pushed a commit, that addresses that. PHAL

**Comment by pracucci** [comment]
Thanks!
